### PR TITLE
Move BLAF.FHA back to child of AS instead of under EJEA.

### PR DESCRIPTION
**Pull request by kshakir**

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala [hunk 1]
```diff
@@ -269,27 +269,15 @@ class EngineJobExecutionActor(replyTo: ActorRef,
       self,
       jobDescriptor,
       initializationData,
-      fileContentsHasherActor,
+      /*
+      Use context.system instead of context as the factory. Otherwise when we die, so will the child actors.
+       */
+      fileContentsHasherSingletonActor(backendName, factory, context.system),
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala]: Can we somehow sneakily use the named `cromwell-root` actor to be the parent? @geoffjentry?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala [hunk 2]
```diff
@@ -444,4 +432,22 @@ object EngineJobExecutionActor {
     * More refinement may appear via #1377.
     */
   private val factorySingletonRouters = scala.collection.mutable.Map[BackendLifecycleActorFactory, ActorRef]()
+
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala]: Think I'd prefer `var` of immutable rather than `val` of mutable

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala (hunk 1 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala]: Discussed in person, answer is no :'(

**Comment by cjllanwarne** [comment]
Otherwise 👍

**Comment by geoffjentry** [comment]
Why do I have a feeling that you're being a bit rebellious with those acronyms?

**Comment by geoffjentry** [comment]
I do think that in practice this is unlikely to matter for a while and this isn't the long term right solution IMO.

That said I also see the logic as to why it was the way it was (even if it was in an incorrect location) 

tl;dr I don't think any of this is a huge deal at the moment so i'm cool with whatever 👍

**Comment by kshakir** [comment]
ToL: I'm fine if these actors hang as children off the initialization actor or some other _actor_, but not as singletons-in-the-system-jvm-referenced-from-the-non-actor-blaf. Relatedly, the name blaf is a slight misnomer, at it actually creates props, not actor refs, so maybe it should be the blpf? Pinging #1377 to record some of this as food for later thought.

**Comment by kshakir** [comment]
Ready for re-review if anyone wants to take a look. Will merge in a couple hours.

File: backend/src/main/scala/cromwell/backend/BackendLifecycleActorFactory.scala [hunk 3]
```diff
@@ -42,7 +42,7 @@ trait BackendLifecycleActorFactory {
   def runtimeAttributeDefinitions(initializationDataOption: Option[BackendInitializationData]): Set[RuntimeAttributeDefinition] = Set.empty
 
   lazy val fileHashingFunction: Option[FileHashingFunction] = None
-  lazy val fileHashingWorkerCount: Int = 50
+  lazy val fileHashingActorCount: Int = 50
 
-  def fileContentsHasherActorProps: Props = FileHasherWorkerActor.props(fileHashingFunction)
+  def fileHashingActorProps: Props = FileHashingActor.props(fileHashingFunction)
```
Comment (cjllanwarne) [inline backend/src/main/scala/cromwell/backend/BackendLifecycleActorFactory.scala]: Info only: 
Oh I remember why `Contents` was in there. I was preparing for other types of file hash (e.g. path, companion `.md5`, etc). I approve of this change

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala [hunk 4]
```diff
@@ -269,27 +269,15 @@ class EngineJobExecutionActor(replyTo: ActorRef,
       self,
       jobDescriptor,
       initializationData,
-      fileContentsHasherActor,
+      /*
+      Use context.system instead of context as the factory. Otherwise when we die, so will the child actors.
+       */
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala]: Mid-argument-list multi-line comments FTL :(

**Comment by geoffjentry** [comment]
:+1:

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala [hunk 5]
```diff
@@ -443,5 +431,54 @@ object EngineJobExecutionActor {
...
+    *
+    * @param backendName                  Name of the backend.
+    * @param backendLifecycleActorFactory A backend factory.
+    * @param actorRefFactory              An actor factory.
+    * @return a RoundRobinPool of actors based on backend factory.
+    */
+  private def factoryFileHashingRouter(backendName: String,
+                                       backendLifecycleActorFactory: BackendLifecycleActorFactory,
+                                       actorRefFactory: ActorRefFactory): ActorRef = {
+    val (originalOrUpdated, result) = getOrElseUpdated(
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala]: ToL: I applaud the FP pureness of this... but couldn't you just:
```
if (!x.contains(y) { x :+= y -> ... }
x(y)
```

**Comment by cjllanwarne** [comment]
Don't let me nit-picking block you: re-👍

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala (hunk 5 above)
Comment (kshakir) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala]: That does an extra lookup. This is an `var immutable.Map` version of the `val mutable.Map.getOrElseUpdate` without the second pass for lookups.

See http://stackoverflow.com/questions/4385976/idiomatic-get-or-else-update-for-immutable-map#answer-5840119
